mem0/llms/gemini.py

```diff
@@ -0,0 +1,139 @@
+from typing import Dict, List, Optional
+
+import google.generativeai as genai
+from google.generativeai import GenerativeModel
+from google.generativeai.types import content_types
+
+from mem0.configs.llms.base import BaseLlmConfig
+from mem0.llms.base import LLMBase
+
+
+class GeminiLLM(LLMBase):
+    def __init__(self, config: Optional[BaseLlmConfig] = None):
+        super().__init__(config)
+
+        if not self.config.model:
+            self.config.model = "gemini-1.5-flash-latest"
+
+        api_key = self.config.api_key or os.getenv("GEMINI_API_KEY")
+        genai.configure(api_key=api_key)
+        self.client = GenerativeModel(model_name=self.config.model)
+
+    def _parse_response(self, response, tools):
+        """
+        Process the response based on whether tools are used or not.
+
+        Args:
+            response: The raw response from API.
+            tools: The list of tools provided in the request.
+
+        Returns:
+            str or dict: The processed response.
+        """
+        if tools:
+            processed_response = {
+                "content": content if (content := response.candidates[0].content.parts[0].text) else None,
+                "tool_calls": [],
+            }
+
+            for part in response.candidates[0].content.parts:
+                if fn := part.function_call:
+                    processed_response["tool_calls"].append(
+                        {
+                                "name": fn.name,
+                                "arguments": {key:val for key, val in fn.args.items()},
+                        }
+                    )
+
+            return processed_response
+        else:
+            return response.candidates[0].content.parts[0].text
+
+    def _reformat_messages(self, messages : List[Dict[str, str]]):
+        """
+        Reformat messages for Gemini.
+
+        Args:
+            messages: The list of messages provided in the request.
+
+        Returns:
+            list: The list of messages in the required format.
+        """
+        new_messages = []
+
+        for message in messages:
+            if message["role"] == "system":
+                content = "THIS IS A SYSTEM PROMPT. YOU MUST OBEY THIS: " + message["content"]
+
+            else:
+                content = message["content"]
+            
+            new_messages.append({"parts": content,
+                                "role": "model" if message["role"] == "model" else "user"})
+
+        return new_messages
+
+    def _reformat_tools(self, tools: Optional[List[Dict]]):
+        """
+        Reformat tools for Gemini.
+
+        Args:
+            tools: The list of tools provided in the request.
+
+        Returns:
+            list: The list of tools in the required format.
+        """
+        if tools:
+            new_tools = []
+
+            for tool in tools:
+                func = tool["function"].copy()
+                func["parameters"].pop("additionalProperties", None)
```

Comment:
Are you sure we don't have to change the underlying tools dict structure here?

Reply:
oh yea I didn't consider nested object properties in the tools dict. In that case, some more changes would need to be made. I'll do that asap.